Here are a few more details about the failures (apparently a certificate issue):

```
2024-11-21T14:03:31.2001912Z [INFO] Running io.quarkus.keycloak.adminclient.deployment.KeycloakAdminClientMutualTlsDevServicesTest
2024-11-21T14:03:34.3108560Z 2024-11-21 14:03:32,234 INFO  [io.sma.cer.CertificateGenerator] (main) ⭐  PKCS12 keystore and truststore generated successfully!
2024-11-21T14:03:34.3144635Z 2024-11-21 14:03:32,234 INFO  [io.sma.cer.CertificateGenerator] (main) 🔐  Key Store File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-keystore.p12
2024-11-21T14:03:34.3149816Z 2024-11-21 14:03:32,235 INFO  [io.sma.cer.CertificateGenerator] (main) 🔓  Server Trust Store File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-server-truststore.p12
2024-11-21T14:03:34.3154144Z 2024-11-21 14:03:32,235 INFO  [io.sma.cer.CertificateGenerator] (main) 🔐  Client Key Store File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-client-keystore.p12
2024-11-21T14:03:34.3158222Z 2024-11-21 14:03:32,235 INFO  [io.sma.cer.CertificateGenerator] (main) 🔓  Client Trust Store File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-client-truststore.p12
2024-11-21T14:03:34.3161335Z 2024-11-21 14:03:32,236 INFO  [io.sma.cer.CertificateGenerator] (main) ⭐  PEM certificates, keystore, and truststore named mtls-test generated!
2024-11-21T14:03:34.3164580Z 2024-11-21 14:03:32,237 INFO  [io.sma.cer.CertificateGenerator] (main) 🔑  Key File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test.key
2024-11-21T14:03:34.3167913Z 2024-11-21 14:03:32,237 INFO  [io.sma.cer.CertificateGenerator] (main) 📜  Cert File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test.crt
2024-11-21T14:03:34.3171610Z 2024-11-21 14:03:32,237 INFO  [io.sma.cer.CertificateGenerator] (main) 🔓  Server Trust Store File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-server-ca.crt
2024-11-21T14:03:34.3175306Z 2024-11-21 14:03:32,238 INFO  [io.sma.cer.CertificateGenerator] (main) 🔑  Client Key File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-client.key
2024-11-21T14:03:34.3196238Z 2024-11-21 14:03:32,238 INFO  [io.sma.cer.CertificateGenerator] (main) 📜  Client Cert File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-client.crt
2024-11-21T14:03:34.3200073Z 2024-11-21 14:03:32,238 INFO  [io.sma.cer.CertificateGenerator] (main) 🔓  Client Trust Store File: /home/runner/work/quarkus/quarkus/extensions/keycloak-admin-resteasy-client/deployment/target/certs/mtls-test-client-ca.crt
2024-11-21T14:03:34.3202942Z 2024-11-21 14:03:34,207 INFO  [io.qua.dev.key.KeycloakDevServicesProcessor] (build-8) Using Quarkus powered Keycloak distribution
2024-11-21T14:03:34.3210606Z 2024-11-21 14:03:34,208 INFO  [tc.qua.io/.0.6] (build-8) Creating container for image: quay.io/keycloak/keycloak:25.0.6
2024-11-21T14:03:34.3213003Z 2024-11-21 14:03:34,268 INFO  [tc.qua.io/.0.6] (build-8) Container quay.io/keycloak/keycloak:25.0.6 is starting: e2b5afdb0a72e48680dd6f828e30e14e16d80e2e0c04b07dc5bd459096665348
2024-11-21T14:03:54.0507290Z 2024-11-21 14:03:53,983 INFO  [tc.qua.io/.0.6] (build-8) Container quay.io/keycloak/keycloak:25.0.6 started in PT19.775192644S
2024-11-21T14:03:54.3510495Z 2024-11-21 14:03:54,342 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-1) javax.net.ssl.SSLHandshakeException: Received fatal alert: bad_certificate [Error Occurred After Shutdown]
2024-11-21T14:03:56.4536646Z 2024-11-21 14:03:56,388 ERROR [io.ver.cor.net.imp.ConnectionBase] (vert.x-eventloop-thread-2) javax.net.ssl.SSLHandshakeException: Received fatal alert: bad_certificate [Error Occurred After Shutdown]
2024-11-21T14:03:58.0553799Z 2024-11-21 14:03:57,989 ERROR [io.qua.dev.key.KeycloakDevServicesProcessor] (build-8) Admin token can not be acquired due to a client connection timeout. You may try increasing the `quarkus.oidc.devui.web-client-timeout` property. [Error Occurred After Shutdown]
2024-11-21T14:03:58.0557297Z 2024-11-21 14:03:57,990 INFO  [io.qua.dev.key.KeycloakDevServicesProcessor] (build-8) Dev Services for Keycloak started.
2024-11-21T14:03:59.4017061Z 2024-11-21 14:03:59,310 WARN  [io.qua.oid.com.run.OidcCommonUtils] (vert.x-eventloop-thread-1) OIDC Server is not available:: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
2024-11-21T14:03:59.4020844Z 	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
2024-11-21T14:03:59.4021969Z 	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:383)
2024-11-21T14:03:59.4023268Z 	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:326)
2024-11-21T14:03:59.4024570Z 	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:321)
2024-11-21T14:03:59.4026190Z 	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1351)
2024-11-21T14:03:59.4028195Z 	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1226)
2024-11-21T14:03:59.4030268Z 	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1169)
2024-11-21T14:03:59.4031718Z 	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:396)
2024-11-21T14:03:59.4033000Z 	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:480)
2024-11-21T14:03:59.4034485Z 	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1277)
2024-11-21T14:03:59.4036097Z 	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1264)
2024-11-21T14:03:59.4037548Z 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
2024-11-21T14:03:59.4038933Z 	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1209)
2024-11-21T14:03:59.4040284Z 	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1679)
2024-11-21T14:03:59.4041453Z 	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1525)
2024-11-21T14:03:59.4042509Z 	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1366)
2024-11-21T14:03:59.4043690Z 	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1415)
2024-11-21T14:03:59.4045175Z 	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530)
2024-11-21T14:03:59.4046865Z 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469)
2024-11-21T14:03:59.4048310Z 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
2024-11-21T14:03:59.4050016Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
2024-11-21T14:03:59.4051923Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
2024-11-21T14:03:59.4053767Z 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
2024-11-21T14:03:59.4055583Z 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
2024-11-21T14:03:59.4057427Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
2024-11-21T14:03:59.4059559Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
2024-11-21T14:03:59.4061173Z 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
2024-11-21T14:03:59.4062772Z 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
2024-11-21T14:03:59.4064496Z 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
2024-11-21T14:03:59.4065957Z 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
2024-11-21T14:03:59.4067419Z 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
2024-11-21T14:03:59.4068627Z 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
2024-11-21T14:03:59.4070118Z 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
2024-11-21T14:03:59.4071552Z 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
2024-11-21T14:03:59.4072902Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2024-11-21T14:03:59.4073968Z 	at java.base/java.lang.Thread.run(Thread.java:840)
2024-11-21T14:03:59.4075984Z Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
2024-11-21T14:03:59.4078275Z 	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439)
2024-11-21T14:03:59.4079700Z 	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306)
2024-11-21T14:03:59.4081287Z 	at java.base/sun.security.validator.Validator.validate(Validator.java:264)
2024-11-21T14:03:59.4082666Z 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:285)
2024-11-21T14:03:59.4084323Z 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:144)
2024-11-21T14:03:59.4086181Z 	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1329)
2024-11-21T14:03:59.4087396Z 	... 31 more
2024-11-21T14:03:59.4088620Z Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
2024-11-21T14:03:59.4090527Z 	at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:148)
2024-11-21T14:03:59.4092240Z 	at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:129)
2024-11-21T14:03:59.4093813Z 	at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)
2024-11-21T14:03:59.4095200Z 	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434)
2024-11-21T14:03:59.4096140Z 	... 36 more
```

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -177,7 +177,7 @@
         <mockito.version>5.14.2</mockito.version>
         <jna.version>5.8.0</jna.version><!-- should satisfy both testcontainers and mongodb -->
         <quarkus-security.version>2.2.0</quarkus-security.version>
-        <keycloak.version>25.0.6</keycloak.version>
+        <keycloak-client.version>26.0.3</keycloak-client.version>
         <logstash-gelf.version>1.15.1</logstash-gelf.version>
         <checker-qual.version>3.48.3</checker-qual.version>
         <error-prone-annotations.version>2.36.0</error-prone-annotations.version>
@@ -6112,39 +6112,23 @@
 
             <dependency>
                 <groupId>org.keycloak</groupId>
-                <artifactId>keycloak-core</artifactId>
-                <version>${keycloak.version}</version>
-            </dependency>
-            <dependency>
-                <groupId>org.keycloak</groupId>
-                <artifactId>keycloak-common</artifactId>
-                <version>${keycloak.version}</version>
-            </dependency>
-            <dependency>
-                <groupId>org.keycloak</groupId>
                 <artifactId>keycloak-admin-client</artifactId>
-                <version>${keycloak.version}</version>
-                <exclusions>
-                    <exclusion>
-                        <groupId>org.checkerframework</groupId>
-                        <artifactId>checker-qual</artifactId>
-                    </exclusion>
-                </exclusions>
-            </dependency>
-            <dependency>
-                <groupId>org.keycloak</groupId>
-                <artifactId>keycloak-adapter-spi</artifactId>
-                <version>${keycloak.version}</version>
+                <version>${keycloak-client.version}</version>
             </dependency>
             <dependency>
                 <groupId>org.keycloak</groupId>
                 <artifactId>keycloak-authz-client</artifactId>
-                <version>${keycloak.version}</version>
+                <version>${keycloak-client.version}</version>
             </dependency>
             <dependency>
                 <groupId>org.keycloak</groupId>
                 <artifactId>keycloak-policy-enforcer</artifactId>
-                <version>${keycloak.version}</version>
+                <version>${keycloak-client.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.keycloak</groupId>
+                <artifactId>keycloak-client-common-synced</artifactId>
+                <version>${keycloak-client.version}</version>
             </dependency>
             <dependency>
                 <groupId>io.quarkus</groupId>

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -95,10 +95,9 @@
         <junit4.version>4.13.2</junit4.version>
 
         <!-- The image to use for tests that run Keycloak -->
-        <!-- IMPORTANT: If this is changed you must also update bom/application/pom.xml and KeycloakBuildTimeConfig/DevServicesConfig in quarkus-oidc/deployment to match the version -->
-        <keycloak.version>25.0.6</keycloak.version>
+        <keycloak.server.version>25.0.6</keycloak.server.version>
         <keycloak.wildfly.version>19.0.3</keycloak.wildfly.version>
-        <keycloak.docker.image>quay.io/keycloak/keycloak:${keycloak.version}</keycloak.docker.image>
+        <keycloak.docker.image>quay.io/keycloak/keycloak:${keycloak.server.version}</keycloak.docker.image>
         <keycloak.docker.legacy.image>quay.io/keycloak/keycloak:${keycloak.wildfly.version}-legacy</keycloak.docker.legacy.image>
 
         <unboundid-ldap.version>7.0.2</unboundid-ldap.version>

--- a/extensions/devservices/keycloak/pom.xml
+++ b/extensions/devservices/keycloak/pom.xml
@@ -34,13 +34,7 @@
         </dependency>
         <dependency>
             <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
+            <artifactId>keycloak-client-common-synced</artifactId>
         </dependency>
     </dependencies>
     <build>

--- a/extensions/keycloak-admin-rest-client/runtime/pom.xml
+++ b/extensions/keycloak-admin-rest-client/runtime/pom.xml
@@ -23,16 +23,6 @@
             <artifactId>quarkus-tls-registry</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>

--- a/extensions/keycloak-admin-rest-client/runtime/src/main/java/io/quarkus/keycloak/admin/client/reactive/runtime/ResteasyReactiveClientProvider.java
+++ b/extensions/keycloak-admin-rest-client/runtime/src/main/java/io/quarkus/keycloak/admin/client/reactive/runtime/ResteasyReactiveClientProvider.java
@@ -20,6 +20,8 @@ import org.jboss.resteasy.reactive.client.impl.WebTargetImpl;
 import org.jboss.resteasy.reactive.server.jackson.JacksonBasicMessageBodyReader;
 import org.keycloak.admin.client.spi.ResteasyClientProvider;
 
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 import io.quarkus.arc.Arc;
@@ -93,7 +95,7 @@ public class ResteasyReactiveClientProvider implements ResteasyClientProvider {
                     clientBuilder = clientBuilder.register(new ClientJacksonMessageBodyWriter(objectMapper));
                 }
             } else {
-                ObjectMapper newObjectMapper = new ObjectMapper();
+                ObjectMapper newObjectMapper = newKeycloakAdminClientObjectMapper();
                 clientBuilder = clientBuilder
                         .registerMessageBodyReader(new JacksonBasicMessageBodyReader(newObjectMapper), Object.class,
                                 HANDLED_MEDIA_TYPES, true,
@@ -136,9 +138,19 @@ public class ResteasyReactiveClientProvider implements ResteasyClientProvider {
     private ObjectMapper getObjectMapper(ObjectMapper value,
             InstanceHandle<ObjectMapper> objectMapperInstance) {
         if (value == null) {
-            return objectMapperInstance.isAvailable() ? objectMapperInstance.get() : new ObjectMapper();
+            return objectMapperInstance.isAvailable() ? objectMapperInstance.get() : newKeycloakAdminClientObjectMapper();
         }
         return value;
+    }
+
+    // creates new ObjectMapper compatible with Keycloak Admin Client
+    private ObjectMapper newKeycloakAdminClientObjectMapper() {
+        ObjectMapper objectMapper = new ObjectMapper();
+        // Same like JSONSerialization class. Makes it possible to use admin-client against older versions of Keycloak server where the properties on representations might be different
+        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        // The client must work with the newer versions of Keycloak server, which might contain the JSON fields not yet known by the client. So unknown fields will be ignored.
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        return objectMapper;
     }
 
     @Override

--- a/extensions/keycloak-admin-rest-client/runtime/src/main/java/io/quarkus/keycloak/admin/client/reactive/runtime/ResteasyReactiveClientProvider.java
+++ b/extensions/keycloak-admin-rest-client/runtime/src/main/java/io/quarkus/keycloak/admin/client/reactive/runtime/ResteasyReactiveClientProvider.java
@@ -6,13 +6,11 @@ import java.util.Optional;
 
 import javax.net.ssl.SSLContext;
 
-import jakarta.enterprise.inject.Instance;
 import jakarta.ws.rs.Priorities;
 import jakarta.ws.rs.client.Client;
 import jakarta.ws.rs.client.WebTarget;
 import jakarta.ws.rs.core.MediaType;
 
-import org.eclipse.microprofile.config.ConfigProvider;
 import org.jboss.resteasy.reactive.client.TlsConfig;
 import org.jboss.resteasy.reactive.client.api.ClientLogger;
 import org.jboss.resteasy.reactive.client.impl.ClientBuilderImpl;
@@ -27,7 +25,6 @@ import com.fasterxml.jackson.databind.ObjectMapper;
 import io.quarkus.arc.Arc;
 import io.quarkus.arc.ArcContainer;
 import io.quarkus.arc.InstanceHandle;
-import io.quarkus.jackson.ObjectMapperCustomizer;
 import io.quarkus.rest.client.reactive.jackson.runtime.serialisers.ClientJacksonMessageBodyWriter;
 import io.quarkus.tls.TlsConfiguration;
 import io.vertx.core.net.KeyCertOptions;
@@ -71,76 +68,19 @@ public class ResteasyReactiveClientProvider implements ResteasyClientProvider {
         if (arcContainer == null) {
             throw new IllegalStateException(this.getClass().getName() + " should only be used in a Quarkus application");
         } else {
-            InstanceHandle<ObjectMapper> objectMapperInstance = arcContainer.instance(ObjectMapper.class);
-            boolean canReuseObjectMapper = canReuseObjectMapper(objectMapperInstance, arcContainer);
-            if (canReuseObjectMapper) {
-
-                ObjectMapper objectMapper = null;
-
-                InstanceHandle<JacksonBasicMessageBodyReader> readerInstance = arcContainer
-                        .instance(JacksonBasicMessageBodyReader.class);
-                if (readerInstance.isAvailable()) {
-                    clientBuilder = clientBuilder.register(readerInstance.get());
-                } else {
-                    objectMapper = getObjectMapper(objectMapper, objectMapperInstance);
-                    clientBuilder = clientBuilder.register(new JacksonBasicMessageBodyReader(objectMapper));
-                }
-
-                InstanceHandle<ClientJacksonMessageBodyWriter> writerInstance = arcContainer
-                        .instance(ClientJacksonMessageBodyWriter.class);
-                if (writerInstance.isAvailable()) {
-                    clientBuilder = clientBuilder.register(writerInstance.get());
-                } else {
-                    objectMapper = getObjectMapper(objectMapper, objectMapperInstance);
-                    clientBuilder = clientBuilder.register(new ClientJacksonMessageBodyWriter(objectMapper));
-                }
-            } else {
-                ObjectMapper newObjectMapper = newKeycloakAdminClientObjectMapper();
-                clientBuilder = clientBuilder
-                        .registerMessageBodyReader(new JacksonBasicMessageBodyReader(newObjectMapper), Object.class,
-                                HANDLED_MEDIA_TYPES, true,
-                                READER_PROVIDER_PRIORITY)
-                        .registerMessageBodyWriter(new ClientJacksonMessageBodyWriter(newObjectMapper), Object.class,
-                                HANDLED_MEDIA_TYPES, true, WRITER_PROVIDER_PRIORITY);
-            }
+            ObjectMapper newObjectMapper = newKeycloakAdminClientObjectMapper();
+            clientBuilder = clientBuilder
+                    .registerMessageBodyReader(new JacksonBasicMessageBodyReader(newObjectMapper), Object.class,
+                            HANDLED_MEDIA_TYPES, true,
+                            READER_PROVIDER_PRIORITY)
+                    .registerMessageBodyWriter(new ClientJacksonMessageBodyWriter(newObjectMapper), Object.class,
+                            HANDLED_MEDIA_TYPES, true, WRITER_PROVIDER_PRIORITY);
             InstanceHandle<ClientLogger> clientLogger = arcContainer.instance(ClientLogger.class);
             if (clientLogger.isAvailable()) {
                 clientBuilder.clientLogger(clientLogger.get());
             }
         }
         return clientBuilder;
-    }
-
-    // the idea is to only reuse the ObjectMapper if no known customizations would break Keycloak
-    // TODO: in the future we could also look into checking the ObjectMapper bean itself to see how it has been configured
-    private boolean canReuseObjectMapper(InstanceHandle<ObjectMapper> objectMapperInstance, ArcContainer arcContainer) {
-        if (objectMapperInstance.isAvailable() && !objectMapperInstance.getBean().isDefaultBean()) {
-            // in this case a user provided a completely custom ObjectMapper, so we can't use it
-            return false;
-        }
-
-        Instance<ObjectMapperCustomizer> customizers = arcContainer.beanManager().createInstance()
-                .select(ObjectMapperCustomizer.class);
-        if (!customizers.isUnsatisfied()) {
-            // ObjectMapperCustomizer can make arbitrary changes, so in order to be safe we won't allow reuse
-            return false;
-        }
-        // if any Jackson properties were configured, disallow reuse - this is done in order to provide forward compatibility with new Jackson configuration options
-        for (String propertyName : ConfigProvider.getConfig().getPropertyNames()) {
-            if (propertyName.startsWith("quarkus.jackson")) {
-                return false;
-            }
-        }
-        return true;
-    }
-
-    // the whole idea here is to reuse the ObjectMapper instance
-    private ObjectMapper getObjectMapper(ObjectMapper value,
-            InstanceHandle<ObjectMapper> objectMapperInstance) {
-        if (value == null) {
-            return objectMapperInstance.isAvailable() ? objectMapperInstance.get() : newKeycloakAdminClientObjectMapper();
-        }
-        return value;
     }
 
     // creates new ObjectMapper compatible with Keycloak Admin Client

--- a/extensions/keycloak-admin-resteasy-client/deployment/src/main/java/io/quarkus/keycloak/adminclient/deployment/KeycloakAdminClientProcessor.java
+++ b/extensions/keycloak-admin-resteasy-client/deployment/src/main/java/io/quarkus/keycloak/adminclient/deployment/KeycloakAdminClientProcessor.java
@@ -14,8 +14,6 @@ import org.keycloak.json.StringOrArraySerializer;
 
 import io.quarkus.arc.BeanDestroyer;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
-import io.quarkus.deployment.Capabilities;
-import io.quarkus.deployment.Capability;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.ExecutionTime;
@@ -58,11 +56,8 @@ public class KeycloakAdminClientProcessor {
     @Record(ExecutionTime.RUNTIME_INIT)
     @Produce(ServiceStartBuildItem.class)
     @BuildStep
-    public void integrate(ResteasyKeycloakAdminClientRecorder recorder, Capabilities capabilities,
-            TlsRegistryBuildItem tlsRegistryBuildItem) {
-        boolean areJSONBProvidersPresent = capabilities.isPresent(Capability.RESTEASY_JSON_JSONB)
-                || capabilities.isPresent(Capability.RESTEASY_JSON_JSONB_CLIENT);
-        recorder.setClientProvider(areJSONBProvidersPresent, tlsRegistryBuildItem.registry());
+    public void integrate(ResteasyKeycloakAdminClientRecorder recorder, TlsRegistryBuildItem tlsRegistryBuildItem) {
+        recorder.setClientProvider(tlsRegistryBuildItem.registry());
     }
 
     @Record(ExecutionTime.RUNTIME_INIT)

--- a/extensions/keycloak-admin-resteasy-client/runtime/pom.xml
+++ b/extensions/keycloak-admin-resteasy-client/runtime/pom.xml
@@ -31,16 +31,6 @@
             <artifactId>quarkus-resteasy-client-jaxb</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>

--- a/extensions/keycloak-admin-resteasy-client/runtime/src/main/java/io/quarkus/keycloak/adminclient/ResteasyKeycloakAdminClientRecorder.java
+++ b/extensions/keycloak-admin-resteasy-client/runtime/src/main/java/io/quarkus/keycloak/adminclient/ResteasyKeycloakAdminClientRecorder.java
@@ -18,6 +18,10 @@ import org.keycloak.admin.client.Keycloak;
 import org.keycloak.admin.client.KeycloakBuilder;
 import org.keycloak.admin.client.spi.ResteasyClientProvider;
 
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
 import io.quarkus.keycloak.admin.client.common.KeycloakAdminClientConfig;
 import io.quarkus.resteasy.common.runtime.jackson.QuarkusJacksonSerializer;
 import io.quarkus.runtime.RuntimeValue;
@@ -66,7 +70,7 @@ public class ResteasyKeycloakAdminClientRecorder {
         };
     }
 
-    public void setClientProvider(boolean areJSONBProvidersPresent, Supplier<TlsConfigurationRegistry> registrySupplier) {
+    public void setClientProvider(Supplier<TlsConfigurationRegistry> registrySupplier) {
         var registry = registrySupplier.get();
         var namedTlsConfig = TlsConfiguration.from(registry,
                 keycloakAdminClientConfigRuntimeValue.getValue().tlsConfigurationName()).orElse(null);
@@ -100,12 +104,10 @@ public class ResteasyKeycloakAdminClientRecorder {
                     }
                 }
 
-                // point here is to use default Quarkus providers rather than org.keycloak.admin.client.JacksonProvider
-                // as it doesn't work properly in native mode
-                if (areJSONBProvidersPresent) {
-                    // when both Jackson and JSONB providers are present, we need to ensure Jackson is used
-                    builder.register(new AppJsonQuarkusJacksonSerializer(), 100);
-                }
+                // this ensures we don't customize managed (shared) ObjectMapper available in the CDI container
+                // and that we use QuarkusJacksonSerializer that works in native mode
+                builder.register(new AppJsonQuarkusJacksonSerializer(), 100);
+
                 return builder.build();
             }
 
@@ -126,6 +128,24 @@ public class ResteasyKeycloakAdminClientRecorder {
     // makes media type more specific which ensures that it will be used first
     @Produces(MediaType.APPLICATION_JSON)
     @Consumes(MediaType.APPLICATION_JSON)
-    static class AppJsonQuarkusJacksonSerializer extends QuarkusJacksonSerializer {
+    static final class AppJsonQuarkusJacksonSerializer extends QuarkusJacksonSerializer {
+
+        private final ObjectMapper objectMapper;
+
+        private AppJsonQuarkusJacksonSerializer() {
+            this.objectMapper = new ObjectMapper();
+            // Same like JSONSerialization class. Makes it possible to use admin-client against older
+            // versions of Keycloak server where the properties on representations might be different
+            this.objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+            // The client must work with the newer versions of Keycloak server, which might contain the JSON fields
+            // not yet known by the client. So unknown fields will be ignored.
+            this.objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        }
+
+        @Override
+        public ObjectMapper locateMapper(Class<?> type, MediaType mediaType) {
+            return objectMapper;
+        }
+
     }
 }

--- a/extensions/keycloak-authorization/deployment/src/main/java/io/quarkus/keycloak/pep/deployment/KeycloakReflectionBuildStep.java
+++ b/extensions/keycloak-authorization/deployment/src/main/java/io/quarkus/keycloak/pep/deployment/KeycloakReflectionBuildStep.java
@@ -5,6 +5,8 @@ import org.keycloak.adapters.authorization.cip.HttpClaimInformationPointProvider
 import org.keycloak.adapters.authorization.cip.spi.ClaimInformationPointProviderFactory;
 import org.keycloak.authorization.client.representation.ServerConfiguration;
 import org.keycloak.authorization.client.representation.TokenIntrospectionResponse;
+import org.keycloak.authorization.client.util.crypto.AuthzClientCryptoProvider;
+import org.keycloak.common.crypto.CryptoProvider;
 import org.keycloak.jose.jwk.JSONWebKeySet;
 import org.keycloak.jose.jwk.JWK;
 import org.keycloak.jose.jws.JWSHeader;
@@ -78,6 +80,8 @@ public class KeycloakReflectionBuildStep {
         serviceProvider.produce(new ServiceProviderBuildItem(ClaimInformationPointProviderFactory.class.getName(),
                 HttpClaimInformationPointProviderFactory.class.getName(),
                 ClaimsInformationPointProviderFactory.class.getName()));
+        serviceProvider.produce(new ServiceProviderBuildItem(CryptoProvider.class.getName(),
+                AuthzClientCryptoProvider.class.getName()));
     }
 
     @BuildStep

--- a/extensions/keycloak-authorization/runtime/pom.xml
+++ b/extensions/keycloak-authorization/runtime/pom.xml
@@ -24,26 +24,11 @@
         </dependency>
         <dependency>
             <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
-            <groupId>org.keycloak</groupId>
             <artifactId>keycloak-policy-enforcer</artifactId>
-            <version>${keycloak.version}</version>
         </dependency>
         <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-adapter-spi</artifactId>
         </dependency>
         <dependency>
             <groupId>org.keycloak</groupId>

--- a/extensions/oidc-client-reactive-filter/deployment/pom.xml
+++ b/extensions/oidc-client-reactive-filter/deployment/pom.xml
@@ -45,6 +45,10 @@
                     <groupId>junit</groupId>
                     <artifactId>junit</artifactId>
                 </exclusion>
+                <exclusion>
+                    <groupId>org.jboss.resteasy</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
             </exclusions>
         </dependency>
         <dependency>

--- a/extensions/oidc-client/deployment/pom.xml
+++ b/extensions/oidc-client/deployment/pom.xml
@@ -50,14 +50,8 @@
         </dependency>
         <dependency>
             <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
+            <artifactId>keycloak-client-common-synced</artifactId>
             <scope>test</scope>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
         </dependency>
         <dependency>
             <groupId>org.eclipse.angus</groupId>

--- a/integration-tests/oidc-client-reactive/pom.xml
+++ b/integration-tests/oidc-client-reactive/pom.xml
@@ -20,16 +20,6 @@
 
     <dependencies>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>

--- a/integration-tests/oidc-client/pom.xml
+++ b/integration-tests/oidc-client/pom.xml
@@ -21,20 +21,15 @@
 
     <dependencies>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>
         <!-- test dependencies -->
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5</artifactId>

--- a/integration-tests/oidc-code-flow/pom.xml
+++ b/integration-tests/oidc-code-flow/pom.xml
@@ -53,13 +53,8 @@
         </dependency>
         <dependency>
             <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>org.eclipse.angus</groupId>

--- a/integration-tests/oidc-tenancy/pom.xml
+++ b/integration-tests/oidc-tenancy/pom.xml
@@ -31,21 +31,16 @@
             <artifactId>quarkus-resteasy-jackson</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>
 
         <!-- test dependencies -->
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-test-keycloak-server</artifactId>

--- a/integration-tests/oidc-token-propagation/pom.xml
+++ b/integration-tests/oidc-token-propagation/pom.xml
@@ -19,20 +19,15 @@
             <artifactId>commons-logging-jboss-logging</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>
         <!-- test dependencies -->
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-test-keycloak-server</artifactId>

--- a/integration-tests/oidc/pom.xml
+++ b/integration-tests/oidc/pom.xml
@@ -32,6 +32,11 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>
@@ -40,7 +45,7 @@
             <artifactId>quarkus-test-security-oidc</artifactId>
             <scope>test</scope>
         </dependency>
-        
+
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5</artifactId>
@@ -177,7 +182,7 @@
                         <configuration>
                             <skip>false</skip>
                             <systemPropertyVariables>
-                                <keycloak.version>${keycloak.version}</keycloak.version>
+                                <keycloak.version>${keycloak.server.version}</keycloak.version>
                             </systemPropertyVariables>
                         </configuration>
                     </plugin>
@@ -186,7 +191,7 @@
                         <configuration>
                             <skip>false</skip>
                             <systemPropertyVariables>
-                                <keycloak.version>${keycloak.version}</keycloak.version>
+                                <keycloak.version>${keycloak.server.version}</keycloak.version>
                             </systemPropertyVariables>
                         </configuration>
                     </plugin>

--- a/integration-tests/smallrye-jwt-oidc-webapp/pom.xml
+++ b/integration-tests/smallrye-jwt-oidc-webapp/pom.xml
@@ -29,20 +29,15 @@
             <artifactId>quarkus-oidc-db-token-state-manager</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>
         <!-- test dependencies -->
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5</artifactId>

--- a/integration-tests/smallrye-jwt-token-propagation/pom.xml
+++ b/integration-tests/smallrye-jwt-token-propagation/pom.xml
@@ -20,16 +20,6 @@
 
     <dependencies>
         <dependency>
-            <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>org.eclipse.angus</groupId>
             <artifactId>angus-activation</artifactId>
         </dependency>
@@ -38,6 +28,11 @@
             <artifactId>quarkus-grpc</artifactId>
         </dependency>
         <!-- test dependencies -->
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-client-common-synced</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-test-security-jwt</artifactId>

--- a/test-framework/keycloak-server/pom.xml
+++ b/test-framework/keycloak-server/pom.xml
@@ -15,13 +15,7 @@
     <dependencies>
         <dependency>
             <groupId>org.keycloak</groupId>
-            <artifactId>keycloak-core</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>com.sun.activation</groupId>
-                    <artifactId>jakarta.activation</artifactId>
-                </exclusion>
-            </exclusions>
+            <artifactId>keycloak-client-common-synced</artifactId>
         </dependency>
         <dependency>
             <groupId>org.testcontainers</groupId>

